File: 01_language_comparison/cycle_detection/cycle_detection.scala
import scala.annotation.tailrec
import scala.collection.mutable

object CycleDetection {
  // Graph represented as an adjacency list
  type Graph = Map[Int, List[Int]]
  
  // Detect cycle in a directed graph using functional approach
  def hasCycle(graph: Graph): Boolean = {
    // Get all vertices
    val vertices = graph.keys.toSet ++ graph.values.flatten.toSet
    
    @tailrec
    def dfs(remaining: Set[Int], visited: Set[Int], recStack: Set[Int]): Boolean = {
      if (remaining.isEmpty) false
      else {
        val vertex = remaining.head
        
        def explore(v: Int, vis: Set[Int], recS: Set[Int]): (Boolean, Set[Int]) = {
          if (recS.contains(v)) (true, vis)
          else if (vis.contains(v)) (false, vis)
          else {
            val newVis = vis + v
            val newRecS = recS + v
            
            val neighbors = graph.getOrElse(v, List.empty[Int])
            
            neighbors.foldLeft((false, newVis)) { case ((cycleFound, currentVis), neighbor) =>
              if (cycleFound) (true, currentVis)
              else explore(neighbor, currentVis, newRecS)
            }
          }
        }
        
        val (hasCycleFromVertex, newVisited) = 
          if (visited.contains(vertex)) (false, visited)
          else explore(vertex, visited, Set.empty[Int])
        
        if (hasCycleFromVertex) true
        else dfs(remaining - vertex, newVisited, Set.empty[Int])
      }
    }
    
    dfs(vertices, Set.empty[Int], Set.empty[Int])
  }
  
  def main(args: Array[String]): Unit = {
    val startTime = System.currentTimeMillis()
    
    // Create a graph as a Map of vertex -> List of adjacent vertices
    val graph: Graph = Map(
      0 -> List(1, 2),
      1 -> List(2),
      2 -> List(0, 3),
      3 -> List(3)
    )
    
    val result = if (hasCycle(graph)) "Graph contains cycle" else "Graph doesn't contain cycle"
    println(result)
    
    val endTime = System.currentTimeMillis()
    println(s"Execution time: ${endTime - startTime} ms")
  }
}

File: 01_language_comparison/cycle_detection/cycle_detection.pl
% Define directed edges in the graph
edge(0, 1).
edge(0, 2).
edge(1, 2).
edge(2, 0).
edge(2, 3).
edge(3, 3).

% Check if there's a path from X to Y with visited nodes list to detect cycles
path(X, Y, Visited) :- 
    edge(X, Z),
    Z \= Y,
    \+ member(Z, Visited),
    path(Z, Y, [Z|Visited]).
path(X, Y, _) :- 
    edge(X, Y).

% Check if there's a cycle starting from Node
has_cycle(Node) :-
    edge(Node, Next),
    path(Next, Node, [Next]).

% Check if the graph has any cycle
graph_has_cycle :-
    has_cycle(_),
    write('Graph contains cycle'), nl.
graph_has_cycle :-
    \+ has_cycle(_),
    write('Graph doesn\'t contain cycle'), nl.

% Main execution with timing in nanoseconds (approximate via get_time)
:- 
    get_time(Start),
    graph_has_cycle,
    get_time(End),
    DurationNs is round((End - Start) * 1_000_000_000),
    format('Execution time: ~w ns~n', [DurationNs]).
